src/Components/Web.JS/src/Boot.Web.ts

```diff
@@ -1,25 +1,48 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
 // Currently this only deals with inserting streaming content into the DOM.
 // Later this will be expanded to include:
 //  - Progressive enhancement of navigation and form posting
 //  - Preserving existing DOM elements in all the above
 //  - The capabilities of Boot.Server.ts and Boot.WebAssembly.ts to handle insertion
 //    of interactive components
 
+import { startCircuit } from './Boot.Server.Common';
+import { startWebAssembly } from './Boot.WebAssembly.Common';
 import { shouldAutoStart } from './BootCommon';
+import { Blazor } from './GlobalExports';
+import { WebStartOptions } from './Platform/WebStartOptions';
 import { attachStreamingRenderingListener } from './Rendering/StreamingRendering';
+import { WebAssemblyComponentDescriptor } from './Services/ComponentDescriptorDiscovery';
+import { ServerComponentDescriptor, discoverComponents } from './Services/ComponentDescriptorDiscovery';
 
 let started = false;
 
-async function boot(): Promise<void> {
+async function boot(options?: Partial<WebStartOptions>): Promise<void> {
   if (started) {
     throw new Error('Blazor has already started.');
   }
   started = true;
+  await activateInteractiveComponents(options);
 
   attachStreamingRenderingListener();
 }
 
-window['Blazor'] = { start: boot }; // Temporary API stub until we include interactive features
+async function activateInteractiveComponents(options?: Partial<WebStartOptions>) {
+  const serverComponents = discoverComponents(document, 'server') as ServerComponentDescriptor[];
+  const webAssemblyComponents = discoverComponents(document, 'webassembly') as WebAssemblyComponentDescriptor[];
+
+  if (serverComponents.length) {
+    await startCircuit(options?.circuit, serverComponents);
+  }
+
+  if (webAssemblyComponents.length) {
+    await startWebAssembly(options?.webAssembly, webAssemblyComponents);
+  }
+}
```

Comment:
I would have hoped that we don't do this this way, and that instead we do an import("...") and call start on the result.

That has important implications for bundling, as we don't want to load the JS/Server code if its not used.

Reply:
Splitting the artifacts into independently-loadable ES modules would be a great benefit. It's probably quite a big extra challenge as we have to think hard about how granularly to split things up so we can limit the amount of overlap (code loaded twice) when both Server and WebAssembly turn out to be used, given the efficiency/simplicity tradeoffs.

While it would be great to do this during .NET 8, I expect we can consider it a relatively low priority as we'll still have a fully-functional product without it.

One possible starting point that sidesteps most of the complexity would be something like:

 * `blazor.web.js`: only includes SSR support (streaming and progressively-enhanced nav, etc.)
   * (Some other child module that gets loaded on demand): includes all the interactive pieces for Server and WebAssembly

This would not involve solving many hard problems since most of those are due to the fine-grained overlap between Server and WebAssembly, and yet still delivers on the promise that SSR can be super lightweight.

Either way this PR is already pretty heavy on refactoring so it would probably be advantageous to save that for later. There is already a list of "todo" items associated with this PR description.

Reply:
All that sounds good to me. Let's save this item for a follow-up PR.